proto/substrait/algebra.proto

```diff
@@ -228,6 +228,90 @@ message AggregateRel {
+    // argument types must also match exactly:
+    //
+    //  - Value arguments must be bound using FunctionArgument.value, and
+    //    the expression in that must yield a value of a type that a function
+    //    overload is defined for.
+    //  - Type arguments must be bound using FunctionArgument.type, and a
+    //    function overload must be defined for that type.
+    //  - Enum arguments must be bound using FunctionArgument.enum
+    //    followed by Enum.specified, with a string that case-insensitively
+    //    matches one of the allowed options.
+    repeated FunctionArgument arguments = 9;
+
+    // Options to specify behavior for corner cases, or leave behavior
+    // unspecified if the consumer does not need specific behavior in these
+    // cases.
+    repeated FunctionOption options = 11;
+
+    // Must be set to the return type of the function, exactly as derived
+    // using the declaration in the extension.
+    Type output_type = 7;
+
+    // Describes which part of the window function to perform within the
+    // context of distributed algorithms. Required. Must be set to
+    // INITIAL_TO_RESULT for window functions that are not decomposable.
+    AggregationPhase phase = 6;
+
+    // Specifies whether equivalent records are merged before being aggregated.
+    // Optional, defaults to AGGREGATION_INVOCATION_ALL.
+    AggregateFunction.AggregationInvocation invocation = 10;
+
+    // Defines the record relative to the current record from which the window
+    // extends. The bound is inclusive. If the lower bound indexes a record
+    // greater than the upper bound, TODO (null range/no records passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the start of the partition.
+    Expression.WindowFunction.Bound lower_bound = 5;
+
+    // Defines the record relative to the current record up to which the window
+    // extends. The bound is inclusive. If the upper bound indexes a record
+    // less than the lower bound, TODO (null range/no records passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the end of the partition.
+    Expression.WindowFunction.Bound upper_bound = 4;
+
+    // Defines the bounds type: ROWS, RANGE
+    BoundsType bounds_type = 12;
+  }
+
+  enum BoundsType {
+    BOUNDS_TYPE_UNSPECIFIED = 0;
+    // The lower and upper bound specify how many rows before and after the current row
+    // the window should extend.
+    BOUNDS_TYPE_ROWS = 1;
+    // The lower and upper bound describe a range of values.  The window should include all rows
+    // where the value of the ordering column is greater than or equal to (current_value - lower bound)
+    // and less than or equal to (current_value + upper bound).  This bounds type is only valid if there
+    // is a single ordering column.
+    BOUNDS_TYPE_RANGE = 2;
+  }
```

Comment:
Can we add `BOUNDS_TYPE_GROUP` while we're adding these?

* https://trino.io/docs/current/sql/pattern-recognition-in-window.html
* https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS

I'm also happy to do this as a later change to reduce the scope of this.

Reply:
This can be done separately.